presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java

```diff
@@ -0,0 +1,228 @@
+import static io.airlift.slice.SizeOf.SIZE_OF_DOUBLE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+
+public class GeometryJtsUtils
+{
+    private static final int POINT = 1;
+    private static final int POLYLINE = 3;
+    private static final int POLYGON = 5;
+    private static final int MULTIPOINT = 8;
+    private static final GeometryFactory FACTORY = new GeometryFactory();
+
+    private GeometryJtsUtils() {}
+
+    /**
+     * Deserializes ESRI shape as described in
+     * https://www.esri.com/library/whitepapers/pdfs/shapefile.pdf
+     * into JTS Geometry object.
+     */
+    public static Geometry deserialize(Slice shape)
+    {
+        if (shape == null) {
+            return null;
+        }
+
+        BasicSliceInput input = shape.getInput();
+        input.skip(SIZE_OF_INT); // SRID
+
+        // GeometryCollection: spatialReferenceId|len-of-shape1|bytes-of-shape1|len-of-shape2|bytes-of-shape2...
+        List<Geometry> geometries = new ArrayList<>();
+        while (input.available() > 0) {
+            int length = input.readInt();
+            geometries.add(readGeometry(input.readSlice(length).getInput()));
+        }
+
+        if (geometries.isEmpty()) {
+            return FACTORY.createGeometryCollection();
+        }
+        else if (geometries.size() == 1) {
+            return geometries.get(0);
+        }
+        return FACTORY.createGeometryCollection(geometries.toArray(new Geometry[0]));
+    }
+
+    private static Geometry readGeometry(SliceInput input)
+    {
+        int geometryType = input.readInt();
+        switch (geometryType) {
+            case POINT:
+                return readPoint(input);
+            case POLYLINE:
+                return readPolyline(input);
+            case POLYGON:
+                return readPolygon(input);
+            case MULTIPOINT:
+                return readMultiPoint(input);
+            default:
+                throw new UnsupportedOperationException("Invalid geometry type: " + geometryType);
+        }
+    }
+
+    private static Point readPoint(SliceInput input)
+    {
+        Coordinate coordinates = readCoordinate(input);
+        if (isEsriNaN(coordinates.x) || isEsriNaN(coordinates.y)) {
+            return FACTORY.createPoint();
+        }
+        return FACTORY.createPoint(coordinates);
+    }
+
+    private static Coordinate readCoordinate(SliceInput input)
+    {
+        return new Coordinate(input.readDouble(), input.readDouble());
+    }
+
+    private static Geometry readMultiPoint(SliceInput input)
+    {
+        skipEnvelope(input);
+        int pointCount = input.readInt();
+        Point[] points = new Point[pointCount];
+        for (int i = 0; i < pointCount; i++) {
+            points[i] = readPoint(input);
+        }
+        return FACTORY.createMultiPoint(points);
+    }
+
+    private static Geometry readPolyline(SliceInput input)
+    {
+        skipEnvelope(input);
+        int partCount = input.readInt();
```

Comment:
You don't need to read `pointCount` if `partCount` is 0 (applies to the `readPolygon` method below)

```
int partCount = input.readInt();
if (partCount == 0) {
    return FACTORY.createPolygon();
}
```